### ISSUE TITLE
Aggregation filter is being pushed down in a way it breaks SQL semantics

### ISSUE THREAD

**Issue by gortiz** [apache/pinot#12647]
Right now in Pinot, aggregation filters like `select SUM(1) FILTER (WHERE a = 1) from Table` are pushed down into something like `SELECT SUM(1) FROM Table WHERE a = 1`. That is fine and it is even desired given it allows us to use indexes in this kind of queries.

But the same optimization is also applied to queries with group by like: `select b, SUM(1) FILTER (WHERE a = 1) from Table group by b`, which are translated to `SELECT b, SUM(1) FROM Table WHERE a = 1`. This is incorrect if there are values of `b` which are never assigned to an `a` whose value is `1`. For example, assuming `Table` is:
```
a | b
-----
1 | 0
2 | 0
2 | 1
```

Then the expected result for `select b, SUM(1) FILTER (WHERE a = 1) from Table group by b` is:
```
b | sum
-------
0 | 1
1 | 0
```

but we are actually returning
```
b | sum
-------
0 | 1
```

In order to correctly follow SQL semantics, we should either:
- Disable this optimization when there are group by.
- Push the filter in some way that the rows used as keys are not filtered out if they don't match the aggregation filter.

At shorter term what we could offer is an option or hint to disable this optimization.

PS: This problem affects both single and multi stage
PS2: See https://modern-sql.com/feature/filter

**Comment by bziobrowski** [comment]
Hello @gortiz  .
I've spent some time debugging and found that this filter pushdown happens during construction of operators (FilteredGroupByOperator to be precise). I think fix could either:
- disable AGGREGATE_CASE_TO_FILTER when aggregation is keyed and add rule for rewriting agg(exp) filter (cond) to agg(case when cond then exp end) for keyed aggregations 
- for keyed aggregations, separate aggregation filters from other filters in buildFilteredAggregationInfos() and introduce an intermediate ProjectionOperator that'd implement agg(case when cond then exp end) .
 
I'm not sure if it's possible to fetch all key combinations that are missing from query output due to filter pushdown. I saw key column dictionary being used to estimate key cardinality but I doubt stats are kept on column value combinations. 

Please let me know what you think .

**Comment by gortiz** [comment]
Thanks @bziobrowski for debugging this. Your solution seems interesting, but it is partial. Pinot has 2 different engines. The original one (we call single-stage or V1) uses Calcite just to parse the expression and apply some simplifications while most of the logical to physical transformations are applied eagerly by our own code. Then we have a second engine (called multi-stage or V2) that relay more in Calcite. This second engine ends up calling the single-stage engine in the leaves of the execution tree.

If I remember correctly, this issue was detected when adding AGGREGATE_CASE_TO_FILTER and we found it affects both single and multi stage. As you say it seems that FilteredGroupByOperator is involved. That class is used in single-stage.

Your first suggestion may be useful, but it is limited to multi-stage and it would end up producing inefficient plans. The reason to merge #12643 was to try to use some indexes we cannot use in filtered aggregations. Also, I'm not sure what do you mean with keyed aggregation. Do you mean using `group by k1, k2, etc`? I don't think that is an actual problematic case. Instead the problem arise when using two aggregate functions and only one of them is filtered.

BTW, it looks like this issue was reported twice (see https://github.com/apache/pinot/issues/13749) and I think @Jackie-Jiang already fixed this in https://github.com/apache/pinot/pull/13784.

**Comment by bziobrowski** [comment]
Thanks @gortiz . From what I see the issue is not fixed in latest master . You can see it if you run and compare 

```sql
select AirTime, count(*) filter ( where AirTime < 0 ) from AirlineStats group by AirTime
```
with
```sql
select AirTime, sum( case when AirTime < 0 then 1 else 0 end)  
from airlineStats 
group by AirTime
```

in query console (assuming instance was started with `./bin/pinot-admin.sh QuickStart -type batch`). 
The second query contains all AirTime values, not only those matching filter predicate. 
It's not necessary to include more than one aggregate function to see the effect. In fact, if filter-less aggregate functions are present, they hide the bug. That's because all group by key combinations produced by  filter-less aggregate get merged with some  produced by filtered aggregate.
By keyed aggregation I meant, as you guessed, one with group by key[s]. 

The way I see it, the problem is that aggregation filters get pushed down even when aggregation is keyed (see AggregationFunctionUtils.buildFilteredAggregationInfos()), leading to filtering of both aggregation keys and values. 
It only seems valid when aggregation doesn't have any key. 

You are right about the first approach related to AGGREGATE_CASE_TO_FILTER being a partial solution since issue affects both engines and that rule only applies to V2. My initial idea was to limit AGGREGATE_CASE_TO_FILTER to keyless aggregates and add reversed rule for keyed aggregates with filter, e.g. 

```
select AirTime, count(*) filter ( where AirTime < 0 )
-->
select AirTime, count( case when AirTime < 0 end )
```
but that transformation is problematic to apply correctly for all aggregate functions (multiple parameters, quirks around null handling, etc.) .

Back to the core issue - I think the most promising approach is `option two` :
`for keyed aggregations, separate aggregation filters from other filters in AggregationFunctionUtils.buildFilteredAggregationInfos() and push it to e.g. DefaultGroupByExecutor to execute it after computing keys but before aggregate values.`

**Comment by kmozaid** [comment]
@gortiz Hi, any update on this issue?

**Comment by gortiz** [comment]
Hi @kmozaid! This was fixed in #14211 last week and will be included in the next release.

**Comment by kmozaid** [comment]
> Hi @kmozaid! This was fixed in #14211 last week and will be included in the next release.

Thank you!!
